### PR TITLE
ENH: `__array_function__` for `np.einsum` and `np.block`

### PR DESCRIPTION
**Pull request by shoyer**
xref https://github.com/numpy/numpy/issues/12028

These turned out to be slightly easier than I expected.

### CODE REVIEW

File: numpy/core/shape_base.py [hunk 1]
```diff
@@ -499,7 +499,16 @@ def _block(arrays, max_depth, result_ndim, depth=0):
         return _atleast_nd(arrays, result_ndim)
 
 
-# TODO: support array_function_dispatch
+def _block_dispatcher(arrays):
+    if type(arrays) is list:
```
Comment (hameerabbasi) [inline numpy/core/shape_base.py]: I'm not sure if we'll be breaking backward compatibility by doing this. Maybe `isinstance(arrays, collections.Iterable) and not hasattr(arrays, '__array_function__')`?

**Review by hameerabbasi — APPROVED** [review]
One minor comment here.

File: numpy/core/shape_base.py (hunk 1 above)
Comment (shoyer) [reply numpy/core/shape_base.py]: My reasoning for picking `type(arrays) is list` is that it matches `np.block()`, which hard codes `list` (not even `tuple`). And we know that `list.__array_function__` will never exist.

Comment (hameerabbasi) [reply numpy/core/shape_base.py]: Okay, in that case ignore my comment. I was worried about the case where we might be inadvertently supporting other kinds of containers in the function but not the dispatcher. A comment would be nice, in any case.

Comment (shoyer) [reply numpy/core/shape_base.py]: Yes, a comment is definitely a good idea here. Done!

**Comment by shoyer** [comment]
@mhvk can I bother you to look over this final `array_function_dispatch` PR? :)

**Review by mhvk — APPROVED** [review]
This looks OK, and hence approved, though arguably the one for `block` is sufficiently non-trivial that tests might be useful. But it does look fine.

**Comment by shoyer** [comment]
@mhvk  I added unit tests for the `block` dispatcher.

**Comment by mhvk** [comment]
OK, looks good, merging!

**Comment by shoyer** [comment]
Thanks Marten!

**Pull request merged**